pkg/adminconfig/evaluator_interfaces.md

```diff
@@ -0,0 +1,259 @@
+type Restrictions struct {
+	// Restrictions on clusters used for deployment
+	Clusters []string `json:"clusters,omitempty"`
+	// Restrictions on modules of the type “key”: “value” when the key is a module property (e.g. scope, type) and the value is an allowed value (e.g. asset, plugin)
+	ModuleRestrictions map[string]string `json:"modules,omitempty"`
+}
+
+// Decision is a result of evaluating a configuration policy which satisfies the specified predicates
+type Decision struct {
+	// a decision regarding deployment: True = require, False = forbid, Unknown = allow
+	Deploy corev1.ConditionStatus `json:"deploy,omitempty"`
+	// Deployment restrictions on modules, clusters and additional resources
+	DeploymentRestrictions Restrictions `json:"restrictions,omitempty"`
+	// Descriptions of policies that have been used for evaluation
+	Policy DecisionPolicy `json:"policy,omitempty"`
+}
+
+type DecisionPerCapabilityMap map[api.CapabilityType]Decision
+```
+
+### Functionality (`adminconfig.EvaluatorInterface`)
+
+Any implementation of the config policy evaluator should implement this interface.
+
+```
+// EvaluatorInterface is an interface for config policies' evaluator
+type EvaluatorInterface interface {
+	SetupWithInfrastructureManager(mgr *InfrastructureManager)
+	Evaluate(in *EvaluatorInput) (EvaluatorOutput, error)
+}
+```
+
+`SetupWithInfrastructureManager` attaches `InfrastructureManager` to the evaluator to obtain the `Infrastructure` object.
+`Evaluate` evaluates config policies based on `EvaluatorInput` and returns `EvaluatorOutput`.
+
+## Interface with OPA
+
+Configuration policies are written in Rego language and are evaluated using OPA (Open Policy Agent).
+Interaction between the evaluator and OPA is done using internal OPA golang packages (see https://pkg.go.dev/github.com/open-policy-agent/opa/rego#Rego.Eval)
+
+Information that passes from config policies' evaluator to OPA has three main ingredeients:
+
+- The `data` json object with infrastructure details, such as available clusters, available object stores, etc. Infrastructure is known at the deployment time and is not changed frequently. The `Infrastructure` object is obtained using `InfrastructureManager` that has interfaces and clients that can access various kubernetes resources, such as configmaps and custom resources. 
+
+- `Modules` with policies created by IT administrator. Modules exist at the deployment time and are unlikely to be changed frequently. 
+
+- Dynamic `input` that has the details of the workload, asset metadata and the user request. Inputs are constructed on each reconciliation of FybrikApplication.
+
+Interaction with OPA is done in two steps. First, a `PreparedEvalQuery` is created using the `data` json and compiled `modules`. This should be done upon changes in policies and/or infrastructure details. Then, the query is evaluated on `input` during each time the plotter object needs to be generated.
```

Comment:
PrepareEvalQuery is done once per fybrik deployment or once when a new fybrikapplication is created?

Reply:
Once per deployment, in the future - upon any change in infrastructure + policies.